src/service/video-manager-impl.js

```diff
@@ -0,0 +1,311 @@
+      viewport.onScroll(scrollListener);
+      viewport.onChanged(scrollListener);
+      this.scrollListenerInstalled_ = true;
+    }
+  }
+}
+
+/**
+ * VideoEntry represents an entry in the VideoManager's list.
+ */
+class VideoEntry {
+  /**
+   * @param {!Window} win
+   * @param {!../video-interface.VideoInterface} video
+   */
+  constructor(win, video) {
+
+    /** @package @const {!../video-interface.VideoInterface} */
+    this.video = video;
+
+    /** @private {boolean} */
+    this.loaded_ = false;
+
+    /** @private {boolean} */
+    this.isVisible_ = false;
+
+    /** @const @private {!../../src/service/vsync-impl.Vsync} */
+    this.vsync_ = vsyncFor(win);
+
+    /** @private {boolean} */
+    this.canAutoplay_ = video.element.hasAttribute(VideoAttributes.AUTOPLAY) &&
+        platformSupportsAutoplay(platformFor(win));
+
+    const element = dev().assert(video.element);
+
+    listenOncePromise(element, VideoEvents.CANPLAY)
+      .then(() => this.videoLoaded_());
+
+    // Currently we only register after video player is build.
+    this.videoBuilt_();
+  }
+
+  /**
+   * Called when the video element is built.
+   * @private
+   */
+  videoBuilt_() {
+    if (this.canAutoplay_) {
+      this.autoplayVideoBuilt_();
+    }
+  }
+
+  /**
+   * Called when the video is loaded and can play.
+   * @private
+   */
+  videoLoaded_() {
+    this.loaded_ = true;
+    if (this.isVisible_) {
+      // Handles the case when the video becomes visible before loading
+      this.loadedVideoVisibilityChanged_();
+    }
+  }
+
+  /**
+   * Called when visibility of a video changes.
+   * @private
+   */
+  videoVisibilityChanged_() {
+    if (this.loaded_) {
+      this.loadedVideoVisibilityChanged_();
+    }
+  }
+
+  /**
+   * Only called when visibility of a loaded video changes.
+   * @private
+   */
+  loadedVideoVisibilityChanged_() {
+    if (this.canAutoplay_) {
+      this.autoplayLoadedVideoVisibilityChanged_();
+    }
+  }
+
+  /**
+   * Whether there is a need to monitor visibility of this video.
+   * @return {boolean}
+   * @package
+   */
+  needsVisibilityObserver() {
+    return this.canAutoplay_;
+  }
+
+  /* Autoplay Behaviour */
+
+  /**
+   * Called when an autoplay video is built.
+   * @private
+   */
+  autoplayVideoBuilt_() {
+    this.video.mute();
+
+    // If autoplay video has controls, hide them and only show them on
+    // user-ineraction.
+    if (this.video.element.hasAttribute(VideoAttributes.CONTROLS)) {
+      this.video.hideControls();
+
+      // TODO(aghassemi): This won't work for iframes, needs a transparent shim
+      listenOnce(this.video.element, 'click', () => {
+        this.video.showControls();
+        this.video.unmute();
+      });
+    }
+  }
+
+  /**
+   * Called when visibility of a loaded autoplay video changes.
+   * @private
+   */
+  autoplayLoadedVideoVisibilityChanged_() {
+    if (this.isVisible_) {
+      this.video.play(/*autoplay*/ true);
+    } else {
+      this.video.pause();
```

Comment:
I've been thinking about this and I don't think we should pause a video that has been interacted with. Even if we don't have the "sticky video" feature yet I don't want to stop listening to the recipe just because I scroll down read more ingredients.